### DISCUSSION TITLE
Catch infinite transitions with `always`

### DISCUSSION THREAD

**Discussion by farskid** [statelyai/xstate#2615] [Ideas]
Currently in XState, if you use `always` to transition to the same state in which `always` lives, it will cause an infinite transition.

```js
createMachine({
  initial: 'a',
  states: {
    a: {
      always: 'a'
    }
  }
})
```

[Link to codesandbox](https://codesandbox.io/s/infinite-transition-loop-with-always-ujgps?file=/src/index.tsx)

Following a private discussion with @Andarist , this could be avoided in the code by hooking into `validateArrayifiedTransitions` and rechecking the necessary information.

**Comment by davidkpiano** [comment]
Listing some of the scenarios here:

```js
a: {
  always: 'a'
}

a: {
  always: [
    { target: 'b', cond: 'someCond' },
    { target: 'a' }
  ]
}

a: {
  always: { actions: 'something' } // no target
}

a: {
  always: { cond: 'someCond', actions: 'something' } // no target
}

a: {
  always: [
    { target: 'b', cond: 'someCond' },
    { actions: 'something' }
  ]
}
```
